### PR TITLE
Auto tray start

### LINKED ISSUE #83: Option to start in tray
Keep app in the tray on startup if start in tray is enabled.

### PR DESCRIPTION
**Pull request by berkeyvx**
Fixes #83 

**Proposed Changes**
 -New switch for start on tray
 -If switch enabled, application starts on closed to tray
 
I did not implement autostart on system startup. In my opinion, autostart on system startup and start on tray are different.It is better to have new switch or option for autostart on system startup.

### CODE REVIEW

**Comment by berkeyvx** [comment]
We should only keep Settings.hpp and Settings.cpp changes. Maybe variable names are could be better. Others are redundant for know.

**Review by eneshecan — CHANGES_REQUESTED** [review]
LGTM other than the tricky part about changing closed to tray status in settings. Instead;

- In the main menu, this option should be insensitive and disabled if closed to tray is not active. So, you should manage its state in the code when closed to tray status changes.
- Then you wouldn't have to do anything dirty to keep their state in sync.

File: src/MainWindow.cpp [hunk 1]
```diff
@@ -126,6 +132,10 @@ bool MainWindow::on_delete_event(GdkEventAny* any_event)
     }
     else
     {
+        if(Settings::instance().startAtTray())
+        {
+            Settings::instance().setCloseToTray(true);
+        }
```
[the remark this answers is not in the document]
Comment (berkeyvx) [reply src/MainWindow.cpp]: It was a trick to make visible tray icon while starting application. I changed this in last commit

**Comment by berkeyvx** [comment]
Can you check again ?

File: resource/ui/MainWindow.ui [hunk 2]
```diff
@@ -130,6 +130,44 @@
             <property name="position">2</property>
           </packing>
         </child>
+        <child>
```
Comment (eneshecan) [inline resource/ui/MainWindow.ui:133]: Just checked and found that `Start in Tray` is a better naming for this feature. Could you please change all the occurrences?

File: src/MainWindow.cpp [hunk 3]
```diff
@@ -73,7 +77,8 @@ MainWindow::MainWindow(BaseObjectType* cobject, Glib::RefPtr<Gtk::Builder> const
 
     show_all();
 
-    m_trayIcon.setVisible(Settings::instance().closeToTray());
+    m_trayIcon.setVisible(Settings::instance().closeToTray() || Settings::instance().startAtTray());
```
Comment (eneshecan) [inline src/MainWindow.cpp:80]: No, tray icon's status should not depend on the status of start in tray. Instead, start in tray should read its value from settings and can only be sensitive if closed to tray is active. So; you should set its sensitivity and status to false in onCloseToTray if closed to tray becomes inactive and save the settings in `onStartInTray` as you did. In that case, onStartInTray will be fired naturally and the setting will be saved.

File: src/main.cpp [hunk 4]
```diff
@@ -21,5 +22,17 @@ int main(int argc, char** argv)
         return 1;
     }
 
-    return app.run(*mainWindow);
+    int retCode = 0;
```
Comment (eneshecan) [inline src/main.cpp:25]: Please use `auto`.

File: src/main.cpp [hunk 5]
```diff
@@ -21,5 +22,17 @@ int main(int argc, char** argv)
...
+    if(Settings::instance().startAtTray())
+    {
+        mainWindow->hide();
+        Application::instance().keepAlive();
+        retCode = app.run();
```
Comment (eneshecan) [inline src/main.cpp:30]: Does it work if you don't pass the mainwindow if you decide to show the app later from the tray icon?

File: src/main.cpp [hunk 6]
```diff
@@ -21,5 +22,17 @@ int main(int argc, char** argv)
...
+    }
+    else
+    {
+        retCode = app.run(*mainWindow);
```
Comment (eneshecan) [inline src/main.cpp:34]: If possible run this first without the if/else branch and then do the things in the if branch to make it simpler. Something like this;
```
auto const retCode = app.run(*mainWindow);
if(Settings::instance().startAtTray())
{
    mainWindow->hide();
    Application::instance().keepAlive();
}

return retCode;
```

File: src/main.cpp [hunk 7]
```diff
@@ -21,5 +22,17 @@ int main(int argc, char** argv)
         return 1;
     }
 
-    return app.run(*mainWindow);
+    int retCode = 0;
+    if(Settings::instance().startAtTray())
```
Comment (eneshecan) [inline src/main.cpp:26]: Please put a whitespace after `if` for readability.

File: src/Settings.cpp [hunk 8]
```diff
@@ -132,3 +132,23 @@ bool Settings::headerBar() const
...
+    {
+        enable = m_keyFile.get_boolean(GROUP_GENERAL, "start_at_tray");
+    }
+    catch (Glib::KeyFileError const& error)
+    {
+        std::cerr << "Settings: " << error.what() << std::endl;
+    }
+
+    return enable;
+}
```
Comment (eneshecan) [inline src/Settings.cpp:154]: Empty line is missing at the end of the file.

File: src/Settings.cpp [hunk 9]
```diff
@@ -132,3 +132,23 @@ bool Settings::headerBar() const
...
+{
+    m_keyFile.set_boolean(GROUP_GENERAL, "start_at_tray", enable);
+}
+
+bool Settings::startAtTray() const
+{
+    auto enable = false;
+    try
+    {
+        enable = m_keyFile.get_boolean(GROUP_GENERAL, "start_at_tray");
```
Comment (eneshecan) [inline src/Settings.cpp:146]: `start_in_tray`

File: src/Settings.cpp [hunk 10]
```diff
@@ -132,3 +132,23 @@ bool Settings::headerBar() const
 
     return enable;
 }
+
+void Settings::setStartAtTray(bool enable)
+{
+    m_keyFile.set_boolean(GROUP_GENERAL, "start_at_tray", enable);
```
Comment (eneshecan) [inline src/Settings.cpp:138]: `start_in_tray`

File: src/MainWindow.hpp [hunk 11]
```diff
@@ -23,6 +23,7 @@ class MainWindow
         void onShow();
         void onQuit(bool immediate);
         bool onCloseToTray(bool state);
+        bool onStartAtTray(bool visible);
```
Comment (eneshecan) [inline src/MainWindow.hpp:26]: `onStartInTray`

File: src/MainWindow.cpp [hunk 12]
```diff
@@ -168,6 +173,12 @@ bool MainWindow::onCloseToTray(bool visible)
     return false;
 }
 
+bool MainWindow::onStartAtTray(bool visible)
```
Comment (eneshecan) [inline src/MainWindow.cpp:176]: `onStartInTray`

File: src/Settings.cpp [hunk 13]
```diff
@@ -132,3 +132,23 @@ bool Settings::headerBar() const
 
     return enable;
 }
+
+void Settings::setStartAtTray(bool enable)
```
Comment (eneshecan) [inline src/Settings.cpp:136]: `setStartInTray`

**Review by eneshecan — CHANGES_REQUESTED** [review]

**Comment by eneshecan** [comment]
I did another review. Please check them and make sure that all the things are working as expected.

File: src/MainWindow.cpp (hunk 3 above)
Comment (berkeyvx) [reply src/MainWindow.cpp:80]: `m_trayIcon.setVisible` sets visibility of tray icon. We should make tray icon visible if either closeToTray or startInTray is set to True. I see nothing wrong here.

```cpp
    m_trayIcon.setVisible(Settings::instance().closeToTray() || Settings::instance().startInTray());
    startInTraySwitch->set_state(Settings::instance().startInTray());
    closeToTraySwitch->set_state(Settings::instance().closeToTray());
```
Can you eleborate more what is wrong with these lines?

File: src/main.cpp (hunk 5 above)
Comment (berkeyvx) [reply src/main.cpp:30]: Yes. 

If i am not wrong, MainWindow::onShow adds window to application when called. `Application::instance().add_window(*this);`

File: src/main.cpp (hunk 6 above)
Comment (berkeyvx) [reply src/main.cpp:34]: I have tried this approach. But it didn't work

**Comment by eneshecan** [comment]
Okay,  LGTM. Merging now.

**Comment by eneshecan** [comment]
Thanks for the great work btw!

**Pull request merged**